File: blocking-sockets/src/main/scala/com/cmhteixeira/concurrency/sockets/blocking/staticfileserver/FileServer.scala
package com.cmhteixeira.concurrency.sockets.blocking.staticfileserver
import org.slf4j.LoggerFactory

import java.io.DataInputStream
import java.net.{InetSocketAddress, NetworkInterface, ServerSocket}
import java.nio.file.Path
import scala.jdk.CollectionConverters.CollectionHasAsScala
import scala.util.Try

object FileServer extends App {

  def parseRequestList(dIS: DataInputStream): Try[Path] = ???

  private val log = LoggerFactory.getLogger("BlockingSockets")
  val interfaces = NetworkInterface.networkInterfaces().toList.asScala
  interfaces.foreach(interface => {
    println(s"#####")
    println(s"DisplayName=${interface.getDisplayName}")
    println(s"HardwareAddress=${Option(interface.getHardwareAddress).map(new String(_)).getOrElse("NA")}")
    println(s"isLoopBack=${interface.isLoopback}")
    interface.getInterfaceAddresses.asScala.foreach(ip => {
      println(s"  IP=${ip.getAddress}")
    })
  })

  val serverSocket = new ServerSocket()
  serverSocket.bind(new InetSocketAddress("127.0.0.1", 8080))
  val clientSocket = serverSocket.accept()
  log.info("Connection accepted ...")

  val inputStream = clientSocket.getInputStream
  val dataInputStream = new DataInputStream(inputStream)


  var exit = false
  while (!exit) {
    println("Entering ...")
    println(new String(inputStream.readNBytes(15)))
    exit = true
  }
}

File: blocking-sockets/src/main/java/com/cmhteixeira/java/sockets/blocking/staticfileserver/FileServer.java
package com.cmhteixeira.java.sockets.blocking.staticfileserver;

import com.cmhteixeira.java.http.HttpParsing;
import com.cmhteixeira.java.http.RequestLine;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.*;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Iterator;
import java.util.List;
import org.slf4j.LoggerFactory;

public class FileServer {
  public static void main(String[] args) throws IOException {
    var log = LoggerFactory.getLogger("BlockingSockets");
    List<NetworkInterface> interfaces = List.of();
    try {
      interfaces = NetworkInterface.networkInterfaces().toList();
    } catch (Exception e) {
      System.err.println("There was an error:");
      e.printStackTrace();
      System.exit(1);
    }

    for (NetworkInterface nInterface : interfaces) {
      System.out.println("#####");
      System.out.println("DisplayName = " + nInterface.getDisplayName());
      System.out.println("isLoopBack = " + nInterface.isLoopback());
      for (Iterator<InetAddress> it = nInterface.getInetAddresses().asIterator(); it.hasNext(); ) {
        InetAddress address = it.next();
        System.out.println("  IP = " + address.getHostAddress());
      }
    }

    Path currentDir = Paths.get("").toAbsolutePath();
    var serverSocket = new ServerSocket();
    serverSocket.bind(new InetSocketAddress("127.0.0.1", 8080));
    var clientSocket = serverSocket.accept();
    log.info("Connection accepted ...");

    var inputStream = clientSocket.getInputStream();
    var dataInputStream = new InputStreamReader(inputStream, StandardCharsets.US_ASCII);
    RequestLine requestLine = HttpParsing.parseRequestLine(dataInputStream);
    System.out.println(requestLine);
    Path requested = Paths.get("/").relativize(Paths.get(requestLine.path()));
    boolean isDir = Files.isDirectory(requested);
  }
}
